### PR TITLE
Kernel515 master

### PR DESCRIPTION
**Pull request by prabirpaul**
**What this PR does / why we need it**:
Bring in changes for kernel 5.15 support from px-fuse v2.12.0.

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:
Testing in progress:
FVT Pass (5.15): https://jenkins.pwx.dev.purestorage.com/job/Dev/job/Porx-09/360/
                             https://jenkins.pwx.dev.purestorage.com/job/Dev/job/Porx-08/57/
BVT Pass (5.15): https://jenkins.pwx.dev.purestorage.com/job/Dev/job/Porx-09/374/
                             https://jenkins.pwx.dev.purestorage.com/job/Dev/job/Porx-09/386/
BVT Pass (3.10): https://jenkins.pwx.dev.purestorage.com/job/Dev/job/Porx-08/50/
                             https://jenkins.pwx.dev.purestorage.com/job/Dev/job/Porx-08/54/
BVT Pass (4.20): https://jenkins.pwx.dev.purestorage.com/job/Dev/job/Porx-07/265/

BVT Fail (5.16): https://jenkins.pwx.dev.purestorage.com/job/Dev/job/Porx-09/363/

### CODE REVIEW

File: pxd.c [hunk 1]
```diff
@@ -1601,15 +1669,29 @@ ssize_t pxd_remove(struct fuse_conn *fc, struct pxd_remove_out *remove)
...
 		}
 
-		pxd_free_disk(pxd_dev);
-		device_unregister(&pxd_dev->dev);
+	spin_unlock(&pxd_dev->lock);
 
-		module_put(THIS_MODULE);
-	}
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,15,0)
+	pxd_free_disk(pxd_dev);
```
Comment (sulakshm) [inline pxd.c:1675]: some part of this code conflicts with the recent issue we found. cant do device_add/device_remove under spinlock.

File: pxd.c [hunk 2]
```diff
@@ -1621,6 +1646,39 @@ ssize_t pxd_remove(struct fuse_conn *fc, struct pxd_remove_out *remove)
 	get_device(&pxd_dev->dev);
 	prepare_to_wait(&pxd_dev->remove_wait, &wait, TASK_INTERRUPTIBLE);
 	spin_unlock(&pxd_dev->lock);
+
+	/// perform all below actions on the kernel object if device got exported.
+	if (pxd_dev->exported) {
```
Comment (sulakshm) [inline pxd.c:1651]: i do not quite follow here, why this needs to be done? This entire logic seems is already part of pxd_finish_remove that is scheduled through a workqueue above. Did i missing anything?

File: pxd.c [hunk 3]
```diff
@@ -1212,24 +1206,59 @@ static int pxd_init_disk(struct pxd_device *pxd_dev)
...
+		return -ENOMEM;
 	  }
 #endif /* __PX_BLKMQ__ */
 #endif /* __PXD_BIO_MAKEREQ__ */
 
+	// Disk and queue initialization
+	snprintf(disk->disk_name, sizeof(disk->disk_name),
+		 PXD_DEV"%llu", pxd_dev->dev_id);
+	disk->major = pxd_dev->major;
+	disk->minors = 256;
```
Comment (sulakshm) [inline pxd.c:1251]: @prabir disk->minors = 1 not 256.

File: pxd.c [hunk 4]
```diff
@@ -1212,24 +1206,59 @@ static int pxd_init_disk(struct pxd_device *pxd_dev)
...
+		 PXD_DEV"%llu", pxd_dev->dev_id);
+	disk->major = pxd_dev->major;
+	disk->minors = 1;
+	disk->first_minor = pxd_dev->minor;
+	disk->flags |= GENHD_FL_EXT_DEVT | GENHD_FL_NO_PART_SCAN;
+#ifdef __PXD_BIO_MAKEREQ__
+		disk->fops = get_bd_fpops();
+#else
+		disk->fops = &pxd_bd_ops;
+#endif
```
Comment (sulakshm) [inline pxd.c:1258]: The above will fail compilation in some releases where block_device_operations does not have submit_bio.
Instead only do:
disk->fops = get_bd_fpops() always.

File: pxd.c (hunk 1 above)
Comment (prabirpaul) [reply pxd.c]: done

File: pxd.c (hunk 2 above)
Comment (prabirpaul) [reply pxd.c]: merge issue, fixed.

File: pxd.c (hunk 3 above)
Comment (prabirpaul) [reply pxd.c]: done

File: pxd.c (hunk 4 above)
Comment (prabirpaul) [reply pxd.c]: done

**Review by sulakshm — APPROVED** [review]
LGTM

**Comment by prabirpaul** [comment]
BVTs all good now. 5.16 has a failure not related to this change:
```
make  -C /usr/src/kernels/5.16.14-1.el7.elrepo.x86_64  M=/home/px-fuse modules
make[1]: Entering directory '/usr/src/kernels/5.16.14-1.el7.elrepo.x86_64'
warning: the compiler differs from the one used to build the kernel
  The kernel was built by: gcc (GCC) 9.3.1 20200408 (Red Hat 9.3.1-2)
  You are using:           gcc-7 (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Kernel version 5.16 supports fastpath.
Kernel version 5.16 supports blkmq driver model.
  SHIPPED /home/px-fuse/px_version.o
  CC [M]  /home/px-fuse/pxd.o
  CC [M]  /home/px-fuse/dev.o
  CC [M]  /home/px-fuse/iov_iter.o
  CC [M]  /home/px-fuse/io.o
/home/px-fuse/io.c: In function 'io_prep_rw':
/home/px-fuse/io.c:572:21: error: assignment from incompatible pointer type [-Werror=incompatible-pointer-types]
  kiocb->ki_complete = io_complete_rw;
                     ^
/home/px-fuse/io.c: In function 'io_rw_done':
/home/px-fuse/io.c:591:3: error: too many arguments to function 'kiocb->ki_complete'
   kiocb->ki_complete(kiocb, ret, 0);
   ^~~~~
/home/px-fuse/io.c:594:3: error: too many arguments to function 'kiocb->ki_complete'
```
This will be addressed separately.

**Pull request merged**